python.mk

```diff
@@ -37,107 +40,132 @@ ${PYTHON} ${ENV_DIR} ${PIP}:
-
+# Builds the wheels for this component
+$${$1_WHEELHOUSE}/.build-wheels: $${$1_SETUP} $${$1_SOURCES} $${$1_WHEELHOUSE}/.clean-wheels
+	# For each setup.py file, generate the wheel
+	$$(foreach setup, $${$1_SETUP}, \
+		$${$1_FLAGS} ${PYTHON} $${setup} $$(if $${$1_BUILD_DIRS},build -b ${ROOT}/build_$$(subst .py,,$${setup})) bdist_wheel -d $${$1_WHEELHOUSE} &&) true
 	touch $$@
 
+${ENV_DIR}/.$1-install-wheels: $${$1_WHEELHOUSE}/.download-wheels
+  # Install all wheels in the wheelhouse into the virtual env for this component
+	${INSTALLER} --find-links=$${$1_WHEELHOUSE} $$(if $${$1_EXTRA_LINKS},--find-links=$${$1_EXTRA_LINKS},) $$(foreach req,$${$1_REQUIREMENTS},-r $${req})
+	touch $$@
 
-endef
+${ENV_DIR}/.$1-install-built-wheels: $${$1_WHEELHOUSE}/.download-wheels $${$1_WHEELHOUSE}/.build-wheels
+  # Install all wheels in the wheelhouse into the virtual env for this component
+	${INSTALLER} --find-links=$${$1_WHEELHOUSE} $$(if $${$1_EXTRA_LINKS},--find-links=$${$1_EXTRA_LINKS},) $${$1_WHEELS}
+	touch $$@
 
-# Common test and coverage targets.
-# To use these, the following must be defined:
-#     * TEST_SETUP_PY: list of the setup.py files that are run under coverage
-#
-# The following are optional, but will be used if defined:
-#     * TEST_PYTHON_PATH: the PYTHONPATH used for the tests
-#     * COMPILER_FLAGS: compiler flags to be used
-#     * COVERAGE_SRC_DIR: the source directory for the coverage run
-#     * COVERAGE_EXCL: excluded files
-#
-.PHONY: coverage
-coverage: ${COVERAGE} ${ENV_DIR}/.test-requirements ${TEST_SETUP_PY}
-	rm -rf htmlcov/
-	${COVERAGE} erase
-	# For each setup.py file in TEST_SETUP_PY, run under coverage
-	$(foreach setup, ${TEST_SETUP_PY}, \
-		$(if ${TEST_PYTHON_PATH},PYTHONPATH=${TEST_PYTHON_PATH},) ${COMPILER_FLAGS} ${COVERAGE} run $(if ${COVERAGE_SRC_DIR},--source ${COVERAGE_SRC_DIR},) $(if ${COVERAGE_EXCL},--omit "${COVERAGE_EXCL}",) -a ${setup} test &&) true
-	${COVERAGE} combine
-	${COVERAGE} report -m --fail-under 100
-	${COVERAGE} html
+# To run the tests, we need to install the dependencies
+${ENV_DIR}/.$1-test-requirements: ${ENV_DIR}/.$1-install-wheels ${ENV_DIR}/.$1-install-built-wheels
 
-.PHONY: test
-test: ${ENV_DIR}/.test-requirements ${TEST_SETUP_PY}
-	# Run test for each setup.py file in TEST_SETUP_PY
-	$(foreach setup, ${TEST_SETUP_PY}, \
-		$(if ${TEST_PYTHON_PATH},PYTHONPATH=${TEST_PYTHON_PATH},) ${COMPILER_FLAGS} ${PYTHON} ${setup} test -v &&) true
-
-# Common test requirements.
-# To use this, the following should be set:
-#     * TEST_REQUIREMENTS: a list of the requirements files needed to install
-#                          the test dependencies
-${ENV_DIR}/.test-requirements: ${ENV_DIR}/.wheels-installed ${TEST_REQUIREMENTS}
-	# Install the test requirements
-	$(foreach reqs, ${TEST_REQUIREMENTS}, \
-		${PIP} install -r ${reqs} &&) true
-	touch $@
+endef
 
-${COVERAGE}: ${PIP}
+${COVERAGE}: ${ENV_DIR}/.env
 	${PIP} install coverage==4.1
```

Comment:
Should we just always install this (and bandit and flake8) as part of the environment?

Reply:
Potentially, but I slightly prefer that you only install them if you need them